### PR TITLE
account/abi: implements event parsing

### PR DESCRIPTION
**Pull request by obscuren**
Implementation of basic event parsing and its input types. This
separates methods and events and fixes an issue with go type parsing and
validation.

### CODE REVIEW

**Comment by robotally** [comment]
| Vote | Count | Reviewers |
| :---: | :---: | :---: |
| :+1: | 0 |  |
| :-1: | 0 |  |

_Updated: Mon Feb 1 15:29:58 UTC 2016_

**Comment by zsfelfoldi** [comment]
Could add a doc comment to func (e Event) Id(), otherwise :+1:

File: accounts/abi/event.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+)
+
+// Event is an event potentially triggered by the EVM's LOG mechanism. The Event
+// holds type information (inputs) about the yielded output
+type Event struct {
+	Name   string
+	Inputs []Argument
+}
+
+func (e Event) Id() common.Hash {
```
Comment (karalabe) [inline accounts/abi/event.go]: Please add a method doc.

File: accounts/abi/abi.go [hunk 2]
```diff
@@ -149,14 +150,36 @@ func (abi ABI) Call(executer Executer, name string, args ...interface{}) interfa
 }
 
 func (abi *ABI) UnmarshalJSON(data []byte) error {
-	var methods []Method
-	if err := json.Unmarshal(data, &methods); err != nil {
+	var fields []struct {
```
Comment (karalabe) [inline accounts/abi/abi.go]: Out of curiosity, shouldn't these field be json tagged?

File: accounts/abi/abi.go [hunk 3]
```diff
@@ -149,14 +150,36 @@ func (abi ABI) Call(executer Executer, name string, args ...interface{}) interfa
...
 		return err
 	}
 
 	abi.Methods = make(map[string]Method)
-	for _, method := range methods {
-		abi.Methods[method.Name] = method
+	abi.Events = make(map[string]Event)
+	for _, field := range fields {
+		switch field.Type {
+		case "function":
```
Comment (karalabe) [inline accounts/abi/abi.go]: According to https://github.com/ethereum/wiki/wiki/Ethereum-Contract-ABI#json, we can have a `constructor` field too and they might also be fully omitted, defaulting to `function`.

Comment (obscuren) [inline accounts/abi/abi.go]: For now we won't support the constructor I guess. I'll add an empty case defaulting to `function`.

Comment (karalabe) [inline accounts/abi/abi.go]: Just wanted to mention it. Alternatively we could also add a fallback case which just prints a warning log so it doesn't silently fail if someone's using it.

Comment (karalabe) [inline accounts/abi/abi.go]: Also please add a test with an empty type to catch this case in the future if we modify something :)

File: accounts/abi/abi.go (hunk 2 above)
Comment (obscuren) [inline accounts/abi/abi.go]: Hmm, no, why?

**Comment by karalabe** [comment]
LGTM :+1:

**Pull request merged**